src/NServiceBus.Core/Unicast/Callback.cs

```diff
@@ -58,60 +60,68 @@ public string MessageId
-            if (!typeof (T).IsEnum)
+            if (!typeof(T).IsEnum)
                 throw new InvalidOperationException(
                     "Register<T> can only be used with enumerations, use Register() to return an integer instead");
 
-            var asyncResult = Register(null, null);
-            var task = Task<T>.Factory.FromAsync(asyncResult, x =>
-                {
-                    var cr = ((CompletionResult) x.AsyncState);
-
-                    return (T) Enum.Parse(typeof (T), cr.ErrorCode.ToString(CultureInfo.InvariantCulture));
-                }, TaskCreationOptions.None, TaskScheduler.Default);
-
-            return task;
+            return InternalRegister(null, null)
+                .ContinueWith(t => (T)Enum.Parse(typeof(T), t.Result.ErrorCode.ToString(CultureInfo.InvariantCulture)),
+                TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
         public Task<T> Register<T>(Func<CompletionResult, T> completion)
         {
-            var asyncResult = Register(null, null);
-            return Task<T>.Factory.FromAsync(asyncResult, x => completion((CompletionResult) x.AsyncState),
-                                                 TaskCreationOptions.None, TaskScheduler.Default);
+            return InternalRegister(null, null).ContinueWith(t => completion(t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
         public Task Register(Action<CompletionResult> completion)
         {
-            var asyncResult = Register(null, null);
-            return Task.Factory.FromAsync(asyncResult, x => completion((CompletionResult) x.AsyncState),
-                                              TaskCreationOptions.None, TaskScheduler.Default);
+            return InternalRegister(null, null).ContinueWith(t => completion(t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
         public IAsyncResult Register(AsyncCallback callback, object state)
         {
+            return InternalRegister(callback, state);
+        }
+
+        Task<CompletionResult> InternalRegister(AsyncCallback callback, object state)
+        {
             if (isSendOnly)
             {
                 throw new Exception("Callbacks are invalid in a sendonly endpoint.");
             }
-            var result = new BusAsyncResult(callback, state);
 
-            if (Registered != null)
-                Registered(this, new BusAsyncResultEventArgs { Result = result, MessageId = messageId });
+            var tcs = new TaskCompletionSource<CompletionResult>(new CompletionResult
+            {
+                State = state
+            });
+
+            if (callback != null)
+            {
+                tcs.Task.ContinueWith(t =>
+                {
+                    try
+                    {
+                        callback(t);
+                    }
+                    catch (Exception e)
+                    {
+                        log.Error(callback.ToString(), e);
```

Comment:
Doesn't `callback.ToString()` gives just a type, in this case `AsyncCallback`, not very useful ?